File: ruby/update.rb
#!/usr/bin/env ruby


# Function definitions
# -----------------------------------------------------------------------------

# Get the split output from the update command
def cmdUpdate
  return %x_sudo apt update 2>/dev/null_.rstrip.split(/\n/)
end

# Check the value of the last line to determine number of packages to upgrade
def checkLast(last)
  return 0 if last === "All packages are up to date."
  match = last.match(/^(?<n>\d+) packages? can be upgraded/)
  return match[:n].to_i if not match.nil?
  raise RuntimeError, "\033[7;31mMessage not recognised:\033[0m\n#{last}"
end

# Ask the user whether to upgrade or not (recursive)
def isUpgrade
  print("\033[7;33mUpgrade packages? [y/n]\033[0m  ")
  is = gets.chomp.downcase
  while not ["y", "n"].include?(is)
    print("\033[7;33mPlease enter only y or n\033[0m\n")
    is = (isUpgrade) ? "y" : "n"
  end
  return is === "y"
end

# Check which packages can be upgraded
def checkUpgradable
  response = %x_apt list --upgradable 2>/dev/null_.split(/\n/)[(1..-1)]
  return response.map{|e| e.match(/^([^\/]+)(?=\/)/)[1]}.sort
end


# Operations
# -----------------------------------------------------------------------------

print("\033[7;34mUpdating package lists....\033[0m\n")

cl = checkLast(cmdUpdate().last)

if cl === 0
  print("\033[7;32mNothing to upgrade\033[0m\n")
  exit 0
end

s = (cl === 1) ? '' : "s"
print("\033[7;33m#{cl} package#{s} to upgrade, fetching list....\033[0m\n")

u = checkUpgradable().join("\n  ")
print("\033[7;34mPackage#{s} to be upgraded:\033[0m\n  #{u}\n")

if isUpgrade()
  print("\033[7;36mUpgrading packages\033[0m\n")
  system("sudo apt -y dist-upgrade")
end

print("\033[7;32mEnd of operations.\033[0m\n")
exit 0

File: julia/update.jl
#!/usr/bin/env julia


# Function definitions
# -----------------------------------------------------------------------------

function cmd_update()::Array{String,1}
  return readlines(pipeline(`sudo apt update`, stderr = devnull))
end

function check_last(last::String)::Union{Nothing,String}
  if last === "All packages are up to date."
    return nothing
  end
  local m = match(r"^(?<n>\d+) packages? can be upgraded", last)
  if isnothing(m)
    println("\e[7;31mMessage not recognised:\e[0m\n$last")
    error("Message not recognised")
  end
  return parse(Int64, m[:n])
end

function is_upgrade()::Bool
  local is::String
  print("\e[7;33mUpgrade packages? [y/n]:\e[0m  ")
  is = lowercase(readline(stdin))
  while !any(x -> x === is, ["y", "n"])
    print("\e[7;35mPlease enter only y or n\e[0m\n")
    is = is_upgrade() ? "y" : "n"
  end
  return is === "y"
end

function check_upgradable()::Array{String,1}
  local resp::Array{String,1}
  resp = readlines(pipeline(`apt list --upgradable`, stderr = devnull))[2:end]
  return sort(map(x::String -> match(r"^([^\/]+)(?=\/)", x), resp))
end


# Operations
# -----------------------------------------------------------------------------

println("\e[7;34mUpdating package lists....\e[0m")

cl = check_last(cmd_update()[end])

if isnothing(cl)
  println("\e[7;32mNothing to upgrade\e[0m")
  exit(0)
end

s = cl === 1 ? '' : "s"
println("\e[7;33m$cl package$s to upgrade, fetching list....\e[0m")

u = join(check_upgradable(), "\n  ")
println("\e[7;34mPackage$s to be upgraded:\e[0m\n  $u")

if is_upgrade()
  println("\e[7;36mUpgrading package$s\e[0m")
  pipeline(`sudo apt -y dist-upgrade`, stderr = devnull, stdout = stdout)
end

println("\e[7;32mEnd of operations\e[0m")
exit(0)
